cranelift/wasm/src/code_translator.rs

```diff
@@ -3579,50 +3581,261 @@ where
+///   increase number of generated instructions and increase cache misses
+fn tag_memory_region_static<FE>(
+    iter_ptr: Value,
+    tagged_ptr: Value,
+    size: u64,
+    size_dynamic: Value,
+    builder: &mut FunctionBuilder,
+    environ: &mut FE,
+) -> WasmResult<()>
+where
+    FE: FuncEnvironment + ?Sized,
+{
+    // Runtime code that is generated:
+    //
+    // (compile-time) num_required_st2g_loops = floor(size / LOOP_UNROLL_BYTES_THRESHHOLD)
+    // (compile-time) num_required_last_stg = (size % LOOP_UNROLL_BYTES_THRESHHOLD) / 16
+    //
+    // (compile-time) if num_required_st2g_loops <= 1
+    // (compile-time)     generate WITHOUT_LOOP_UNROLLING_THRESHHOLD
+    // (compile-time) else
+    // (compile-time)     generate WITH_LOOP_UNROLLING_THRESHHOLD
+    //
+    // WITHOUT_LOOP_UNROLLING_THRESHHOLD:
+    // +---less_than_threshhold
+    // | st2g(tagged_ptr, iter_ptr + 0)          \
+    // | st2g(tagged_ptr, iter_ptr + 32)          | repeat `n` times, i.e. while (remaining_bits >= 32) holds
+    // | ...                                      |
+    // | st2g(tagged_ptr, iter_ptr + (n-1) * 32) /
+    // |
+    // | stg(tagged_ptr, iter_ptr + n * 32)  | generate only if required
+    // +---
+    //
+    // WITH_LOOP_UNROLLING_THRESHHOLD:
+    // end_ptr = iter_ptr + size
+    // +---st2g_loop_condition_block(iter_ptr)
+    // | if (iter_ptr < end_ptr)
+    // |     goto st2g_loop_body_block(iter_ptr)
+    // | else
+    // |     (compile-time) if (num_required_last_stgs == 0)
+    // |                        goto next_block
+    // |     (compile-time) else
+    // |                        goto last_stgs_block
+    // |
+    // | +---st2g_loop_body_block(iter_ptr, st2g_loop_counter)
+    // | | st2g(tagged_ptr, iter_ptr + 0)          \
+    // | | st2g(tagged_ptr, iter_ptr + 32)          | repeat `n`=LOOP_UNROLL_BYTES_THRESHHOLD times
+    // | | ...                                      |
+    // | | st2g(tagged_ptr, iter_ptr + (n-1) * 32) /
+    // | |
+    // | | iter_ptr += LOOP_UNROLL_BYTES_THRESHHOLD
+    // | | goto st2g_loop_condition_block(iter_ptr)
+    // | +---
+    // +---
+    //
+    // +---last_stgs_block(iter_ptr)
+    // | st2g(tagged_ptr, iter_ptr + 0)          \
+    // | st2g(tagged_ptr, iter_ptr + 32)          | repeat `n` times, i.e. while (remaining_bits >= 32) holds
+    // | ...                                      |
+    // | st2g(tagged_ptr, iter_ptr + (n-1) * 32) /
+    // |
+    // | stg(tagged_ptr, iter_ptr + n * 32)  | generate only if required
+    // +---
+    //
+    // +---next_block
+    // +---
+
+    // If the number of bytes to be tagged is larger than this threshhold,
+    // insert a loop that tags using `st2g` chunks.
+    // Threshhold should be a multiple of 32, so each `st2g` chunk can be
+    // exactly divided into only 32-byte tagging `st2g` instructions
+    // This constant was adapted from clang.
+    const LOOP_UNROLL_BYTES_THRESHHOLD: u32 = 160;
```

Comment:
I think making this a `u32` instead of `u64` makes more sense. I wanted to add some safer conversions from `u64` to `i64` where possible (I know that that conversion might only lead to a failure if the `u64` is `u64::MAX`, which `LOOP_UNROLL_BYTES_THRESHHOLD` obviously isn't, but I feel safer adding the explicit conversions with `u64::from` over casts nonetheless). The reason I chose `u32` is because it can always be converted to a `i64` (i.e. `i64` implements the `From<u32>` trait, but not the `From<u64` trait). In some other areas I have used `i64::try_from` or similar to rather give a wasmtime/cranelift error than to risk an invalid conversion.